datalad_next/utils/shell_connection.py

```diff
@@ -0,0 +1,305 @@
+        stdin : Iterable[byte] | None, optional, default: None
+            If given, the bytes are sent to the stdin of the command.
+
+            If the
+            command reads its ``stdin`` until EOF, you have to use
+            :meth:`self.close` to close ``stdin`` of the command. Otherwise
+            the command will usually not terminate.
+            Once :meth:`self.close` is called, no more commands can be executed
+            with this :class:`ShellCommandExecutor`-instance.
+
+        Yields
+        ------
+        :class:`ShellCommandResponseGenerator`
+
+            A generator that yields the output of ``stdout`` of the command.
+            The generator is exhausted when all output is read. After that,
+            the return code of the command execution is available in the
+            ``code``-attribute of the generator.
+        """
+        self.process_inputs.put([command + self.command_postfix])
+        if stdin is not None:
+            self.process_inputs.put(stdin)
+        return ShellCommandResponseGenerator(self.stdout, self.end_marker)
+
+    def upload(self,
+               local_path: Path,
+               remote_path: PurePosixPath,
+               ) -> tuple[int, bytes]:
+        """Upload a local file to a named file in the connected shell
+
+        Parameters
+        ----------
+        local_path : Path
+            The file that should be uploaded.
+        remote_path : PurePosixPath
+            The name of the file on the connected shell that will contain the
+            uploaded content.
+
+        Returns
+        -------
+        tuple[int, bytes]
+            A tuple containing the return code of the command that was executed
+            to upload the content and the last 64k bytes of stderr output
+        """
+        file_size = local_path.stat().st_size
+        cmd_line = f'dd bs=1 of="{remote_path.as_posix()}" count={file_size}'
+        with local_path.open('rb') as local_file:
+            result = self.execute_command(
+                cmd_line.encode(),
+                stdin=iter(local_file.read, b''))
+            consume(result)
+        return result.returncode, b''.join(result.stderr_deque)
+
+    def download(self,
+                 remote_path: PurePosixPath,
+                 local_path: Path,
+                 ) -> tuple[int, bytes]:
+        """Download a file from the connected shell
+
+        Parameters
+        ----------
+        remote_path : PurePosixPath
+            The name of the file on the connected shell that should be
+            downloaded.
+        local_path : Path
+            The name of the local file that will contain the downloaded content.
+
+        Returns
+        -------
+        tuple[int, bytes]
+            A tuple containing the return code of the command that was executed
+            to upload the content and the last 64k bytes of stderr output
+        """
+        # We use 7z in gzip-mode to compress the data that is sent over the
```

Comment:
`gzip` does not seem to be "a" mode, but "the" mode. I did not see a switch to turn this off.

Maybe I am not understanding it properly, but what this compression feature not be redundant with `ssh -C`?

Reply:
> `gzip` does not seem to be "a" mode, but "the" mode. I did not see a switch to turn this off.

I will update/remove the comment.

> 
> Maybe I am not understanding it properly, but what this compression feature not be redundant with `ssh -C`?
We use the compression feature mainly to ensure that the end-marker is not present in the data stream that is received from the server, for example, if the downloaded file contains an end-marker.

Reply:
We now use base64 to encode uploaded and downloaded content. This keeps the end-marker significant and prevents the sending of control-characters

Reply:
We don't use `7z` anymore. Instead we use `base64` to keep the end-marker significant and to prevent sending of control sequences.